### PR TITLE
Upate isVisible() to consider Blockly.editMode

### PR DESCRIPTION
**Pull request by Hamms**
Namely, isVisible() will now return true when in edit mode if the block
is not userVisible but is otherwise visible.

Also, connection now uses this method for determining whether or not it
can connect to the block. (previously when it edit mode, it let you
connect to all blocks, including ones that were actually hidden).

### CODE REVIEW

**Comment by Hamms** [comment]
fixes https://www.pivotaltracker.com/n/projects/1256652/stories/103931484

**Comment by Bjvanminnen** [comment]
it would be great if we could get some test coverage of the different block visibility scenarios :)

File: blockly-core/core/ui/block.js [hunk 1]
```diff
@@ -1542,12 +1542,19 @@ Blockly.Block.prototype.setCurrentlyHidden = function (hidden) {
...
  * @returns true if both visibility conditions are met.
  */
 Blockly.Block.prototype.isVisible = function () {
   var visibleThroughParent = !this.parentBlock_ || this.parentBlock_.isVisible();
-  return visibleThroughParent && this.isUserVisible() && !this.isCurrentlyHidden_();
+  var visible = visibleThroughParent && !this.isCurrentlyHidden_();
+
+  if (!Blockly.editBlocks) {
+    visible = visible && this.isUserVisible();
+  }
```
Comment (Bjvanminnen) [inline blockly-core/core/ui/block.js:1557]: Let me see if I'm reading this right. If we're in editBlock mode (i.e. Blockly.editBlocks evaluates to truthy) we report as being visible even if this.isUserVisible evaluates to false

Comment (Bjvanminnen) [inline blockly-core/core/ui/block.js:1557]: I might rewrite this section as

```
if (Blockly.editBlocks) {
  // We want to show this blocks whether or not isUserVisible is true
  return visible;
}

return visible && this.isUserVisible();
```

I thinks it's easier to wrap my head around without the negative.

**Comment by bcjordan** [comment]
LGTM! Only thing to manually check would be double checking course 4 function editor within level building mode works as expected.

It may also now be possible to write a blockly-core test that covers some of these block visibility scenarios at a lower level.

File: blockly-core/core/ui/block.js (hunk 1 above)
Comment (Hamms) [inline blockly-core/core/ui/block.js:1557]: yeah, that makes sense. I'll throw in a comment mentioning that editors aren't considered "users" to make it clearer why we want to show the block in this situation

**Comment by Hamms** [comment]
will do on both counts

**Comment by Hamms** [comment]
@bcjordan @Bjvanminnen I added a couple tests, and all tests (including eyes tests) pass locally; this look good?

**Comment by Bjvanminnen** [comment]
One minor comment on one of the commits, otherwise lgtm

**Pull request merged**